### ISSUE TITLE
Property: numerous problems with registering/deregistering observers

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/axon#129]
`link` and `unlink` are currently no-ops if `observer` is already registered or not registered respectively.  Likely to silently ignore bugs in client code.

`linkWithDisposal` calls `link`, but then adds a listener, regardless of whether `observer` was already registered.

`lazyLink` doesn't check, and registers `observer`. 

`once` has the same problems as `lazyLink`.

`linkAttribute` and `onValue` don't check. They create and register a new listener.

`hasListener` should be called `hasObserver`, because all parameters refer to `observer`.

This may not be an exhaustive list.

Guess I'd better get in the habit of calling `hasListener`.

**Comment by jonathanolson** [comment]
Noticed a potential similar issue with Node's addInputListener today with @samreid. Added assertion in https://github.com/phetsims/scenery/issues/651.

**Comment by samreid** [comment]
I removed `linkWithDisposal`.

**Comment by samreid** [comment]
In my opinion, Emitter (instead of Property) should define the semantics for what should happen when trying to add duplicate listeners or remove non-existent listeners.  Property should leverage Emitter as much as possible, and hence provide the same semantics.

**Comment by samreid** [comment]
> linkAttribute and onValue don't check. They create and register a new listener.

Hence there is no need to check, since the listener is a new one?

**Comment by samreid** [comment]
> Guess I'd better get in the habit of calling `hasListener`.

Let's decide how this should behave and implement it like that.

**Comment by samreid** [comment]
It looks like about half of the simulations are affected by the "duplicate add" or "tried to remove nonexistent" problems:
![image](https://user-images.githubusercontent.com/679486/28603837-da2e51ae-7183-11e7-8e04-fd94d21fd0c3.png)
I recommend to investigate solving all of these.

**Comment by pixelzoom** [comment]
I ran acid-base-solutions with `?ea&fuzzMouse` and I see no problem.  How did you generate the above "problems"?

**Comment by samreid** [comment]
I added assertions in Emitter to prevent multiple registrations of the same listener or removing non-existent listeners.

**Comment by pixelzoom** [comment]
I said:
> linkAttribute and onValue don't check. They create and register a new listener.

@samreid said:
> Hence there is no need to check, since the listener is a new one?

The problem with linkAttribute and onValue  is that they wrap the observer that's provided, and return a new function that you then need to use to "unlink".   So you can redundantly add the same observer.  And when de-registering, there's nothing that prevents you from calling `unlinkLinkAtttribute` with an observer that wasn't created via `linkAttribute`.   The "link attribute" API is convenient, but not at all robust.

**Comment by jonathanolson** [comment]
So is the ideal behavior desired?:

- Only one of a particular listener can be added (or else assertion failure)
- Adding/removing listeners shouldn't no-op (assertion failure if trying to remove a listener that isn't a listener, etc.)

I've generally been operating under those assumptions when adding/removing listeners (sounds good to me), but I don't want to repeat the "I add assertions, everyone has to do work, but we don't gain much of a benefit".

**Comment by pixelzoom** [comment]
Any consistent approach would be better than what we have now.  But my preference would be what @jonathanolson described in https://github.com/phetsims/axon/issues/129#issuecomment-318187907t, because I believed that (a) function calls that silently do nothing are generally a bad idea, and (b) registering/deregistering something twice is either sloppy programming or a programming error.  And yes, that will require fixing many sims that rely on the current no-op behavior.

**Comment by samreid** [comment]
I agree with your reasoning--the only counterexample I can think of is that we decided on a "sloppy" approach for dispose (I don't recall the issue though).

**Comment by pixelzoom** [comment]
Calling `dispose` twice causes an assertion failure.

**Comment by pixelzoom** [comment]
We compromised on "use after disposal", because scuttling an object was causing problems when callbacks that used that object had been scheduled prior to the dispose call.

**Comment by pixelzoom** [comment]
@samreid said:
> I added assertions in Emitter to prevent multiple registrations of the same listener or removing non-existent listeners.

I added these assertions to Emitter `addListener` and `removeListener` respectively:

```js
assert && assert( !this.hasListener( listener ) );
assert && assert( this.hasListener( listener ) );
```

I tested acid-base-solutions and beers-law-lab with `?ea&fuzzMouse`.  I see no problems.  What am I doing wrong?

**Comment by samreid** [comment]
You probably also need to remove the guards in Property.

**Comment by pixelzoom** [comment]
In https://github.com/phetsims/axon/issues/129#issuecomment-318191403, I said:
> ... (b) registering/deregistering something twice is either sloppy programming or a programming error

For example, I investigated the failure in acid-base-solutions.  Indeed, it was a programming error, see https://github.com/phetsims/acid-base-solutions/issues/152.   Not something that would impact the user, but it did put the sim in an invalid intermediate state.

**Comment by samreid** [comment]
Meeting consensus: it would be nice to prevent duplicates and require a listener to be present when added.  This will be done in Emitter.

@jonathanolson suggested adding maps when assertions are enabled.
@samreid will investigate.
@jessegreenberg wants to tag along.

We will see how it goes it after a couple of hours.

**Comment by samreid** [comment]
@jessegreenberg and I looked at this and it seems it would be possible to add something like this to `once`:

```js
      once: function( listener ) {
        var self = this;
        var wrapper = function( newValue, oldValue ) {
          self.unlink( wrapper );
          listener( newValue, oldValue );
        };
        if (assert){
          listener.wrappedListener = listener;
          wrapper.equals = function(other){
            return listener === other.wrappedListener;
          };
        }
        this.lazyLink( wrapper );
        return wrapper;
      },
```

Then we can use reference equality combined with `equals` to check for existence of listeners. 

```js
        // We need to prohibit this too
        myProperty.once( hello );
        myProperty.linkAttribute( 'x', hello );
```

**Comment by jonathanolson** [comment]
What is the preferred method of handling this pattern (comes up in https://github.com/phetsims/scenery/issues/660):

```js
// Imagine you have lots of Properties
var aProperty = new Property( ... );
var bProperty = new Property( ... );
// ...
var zProperty = new Property( ... );

// And you get passed an array of properties with duplicates
var properties = [ a, a, b, h, a, w, i, a, b, w, q, p ];

// And you want to add a listener to see if any of them change:
this.listener = this.someMethod.bind( this );
for ( var i = 0; i < properties.length; i++ ) {
  properties[ i ].lazyLink( this.listener );
}
```

This will fail out when it tries to link the second copy of `a` to the same listener.

Ideal prerequisites for a solution would:

1. Not create temporary objects (GC is an issue here, thus the for-loop and persistent listener)
2. Not be horrible for performance (naive deduplication of `n` items would have `n^2` comparisons).

**Comment by jonathanolson** [comment]
Current best solution:
```js
// before
function add( property ) {
  property.lazyLink( this.listener );
}
function remove( property ) {
  property.unlink( this.listener );
}

// after
var properties = [];
var counts = [];
function add( property ) {
  var index = _.indexOf( properties, property );
  if ( index >= 0 ) {
    counts[ index ]++
  }
  else {
    properties.push( property );
    counts.push( 1 );
    property.link( this.listener );
  }
}
function remove( property ) {
  var index = _.indexOf( properties, property );
  counts[ index ]--;
  if ( counts[ index ] === 0 ) {
    properties.splice( index, 1 );
    counts.splice( index, 1 );
    property.unlink( this.listener );
  }
}
```

**Comment by samreid** [comment]
@jonathanolson and I discussed an issue around link and duplicates, would be good to discuss with @pixelzoom before we proceed: https://github.com/phetsims/scenery/issues/660

**Comment by pixelzoom** [comment]
I looked at phetsims/scenery#660 and asked questions. Reassigning to @samreid.

**Comment by samreid** [comment]
On hold until https://github.com/phetsims/scenery/issues/660 discussion is further along

**Comment by samreid** [comment]
@jonathanolson modified scenery so we could add this line to Emitter.addListener now:

```js
assert && assert( this.listeners.indexOf( listener ) === -1, 'Cannot add the same listener twice' );
```

**Comment by samreid** [comment]
Enough aqua is passing that it seems safe enough to add this emitter check now:

![image](https://user-images.githubusercontent.com/679486/29243203-1daa6090-7f57-11e7-86eb-aeba066d263e.png)

If any issues crop up, we should try to address them.

**Comment by samreid** [comment]
It may be best to proceed in phases.

Phase 1: remove the guard from Property.unlink and update all sims to make sure it is OK
Phase 2: remove the guard from Property.link and update all sims to make it OK
Phase 3: for higher-order Property methods, use a strategy like the one @jessegreenberg and I described in https://github.com/phetsims/axon/issues/129#issuecomment-320079445 and update all sims.

We will likely need help from all sim developers to do these phases.

**Comment by samreid** [comment]
After we investigate Phase 1 we can decide if we want to follow through with it, and proceed with the following phases.  This issue is on hold until we have more intel on Phase 1.

**Comment by oliver-phet** [comment]
We'll leave this marked for next week to see how things are going.

**Comment by jessegreenberg** [comment]
Removing assignment until we come back to this issue after Phase 1.

**Comment by samreid** [comment]
Phase 1 is complete and Phase 2 is nearly complete.  Let's wait until Phase 2 is complete then merge the axon branch to master.

Summarizing status from original post, assuming Phase 1 and Phase 2 complete:

- [ ] `link` and `unlink` are currently no-ops if `observer` is already registered or not registered respectively.  Likely to silently ignore bugs in client code.

- [ ] `linkWithDisposal` calls `link`, but then adds a listener, regardless of whether `observer` was already registered.

- [ ] `lazyLink` doesn't check, and registers `observer`. 

- [ ] `once` has the same problems as `lazyLink`.

- [ ] `linkAttribute` and `onValue` don't check. They create and register a new listener.

- [ ] `hasListener` should be called `hasObserver`, because all parameters refer to `observer`.

**Comment by samreid** [comment]
Regarding the unchecked boxes:

We can use the proposed strategy for dealing with higher-order functions (including once/linkAttribute/onValue/etc), but it should be confirmed by a quorum of @jonathanolson @jbphet @pixelzoom @zepumph before we invest in it (I left off me and @jessegreenberg because we already discussed it).

> `hasListener` should be called `hasObserver`, because all parameters refer to `observer`.

`listen*` occurs 84 times in Property.js whereas `observ*` occurs twice, so this precedent suggests we should eliminate the `observer` terminology and move toward listener.  Sound OK?

**Comment by samreid** [comment]
Another option to deal with the wrapper-style functions would be to eliminate them and require clients to use link/unlink/lazyLink

**Comment by pixelzoom** [comment]
> `listen*` occurs 84 times in Property.js whereas `observ*` occurs twice, so this precedent suggests we should eliminate the `observer` terminology and move toward listener.  Sound OK?

Observer does seem more appropriate for Property.

> Another option to deal with the wrapper-style functions would be to eliminate them and require clients to use link/unlink/lazyLink

 If you mean `once`, `toggle`, `onValue`, etc. ... I don't think I ever used these.  Not sure if that's a comment on their value or my general tendency to avoid API features that I find questionable.

**Comment by jbphet** [comment]
> `listen*` occurs 84 times in Property.js whereas `observ*` occurs twice, so this precedent suggests we should eliminate the `observer` terminology and move toward listener.  Sound OK?

No strong preference - I'm fine with either, but agree that they should be consistent.

> Another option to deal with the wrapper-style functions would be to eliminate them and require clients to use link/unlink/lazyLink

Unlike @pixelzoom, I **have** used features of axon like `once`, `linkAttribute`, and `onValue`, and would find it a bit frustrating if I had to rework previous sims because those API calls were suddenly rescinded because we didn't want to maintain them.  If there are unused APIs, I'd be fine with eliminating them, but I'd vote to keep the ones that are used.

**Comment by samreid** [comment]
Maybe something like this?

```js
property.link(Property.attributeListener(myObject,'color'));
```

**Comment by samreid** [comment]
```js
var listener = Property.attributeListener(myObject,'color');
property.link(listener);
property.unlink(listener);
```

**Comment by samreid** [comment]
It seems we use `linkAttribute` mainly in places where it doesn't need to be unlinked.

**Comment by jonathanolson** [comment]
I have no strong preferences, the current way of doing things (linkAttribute/once/etc.) and the proposed way (creation patterns for listeners, removal) would both be fine with me.

**Comment by jessegreenberg** [comment]
I slightly prefer second option in https://github.com/phetsims/axon/issues/129#issuecomment-334545950, ```linkAttribute``` is meant to make things more readable, and something like ```Property.attributeListener``` negates that benefit.

I don't feel strongly about creating symmetry for ```link``` and ```unlink``` for this

**Comment by samreid** [comment]
I'm somewhat concerned with the cost of iterating over hundreds of uses of the legacy API to convert to an API that doesn't seem to be significantly better.

We will wait for @pixelzoom to discuss this further.

**Comment by samreid** [comment]
> Observer does seem more appropriate for Property.

That seems reasonable to me, though anecdotally I noticed that during developer meeting another developer used the phrase "add a property listener" instead of "add a property observer".

**Comment by pixelzoom** [comment]
What part of this are we going to address at 10/19/2017 developer meeting?

**Comment by samreid** [comment]
https://github.com/phetsims/axon/issues/129#issuecomment-334475313 through https://github.com/phetsims/axon/issues/129#issuecomment-334550201

**Comment by samreid** [comment]
We are going to remove Property.once ( @jbphet ) and Property.onValue (@samreid).  If I need help on onValue, I'll create a chip-away issue.

**Comment by pixelzoom** [comment]
Also... Keeping `linkAttribute`.  Keeping "listener" terminology.

**Comment by samreid** [comment]
Issues were created for removing Property.once and Property.onValue, closing.
